File: android/app/src/main/java/com/sandnes/familyapp/ui/birthday/BirthdayDateUtils.kt
package com.sandnes.familyapp.ui.birthday

import java.time.LocalDate

internal fun nextBirthdayDate(
    isoDate: String,
    today: LocalDate = LocalDate.now(),
): LocalDate? =
    runCatching {
        val bd = LocalDate.parse(isoDate)
        val thisYear = bd.withYear(today.year)
        if (thisYear < today) bd.withYear(today.year + 1) else thisYear
    }.getOrNull()

internal fun turnsAge(
    isoDate: String,
    today: LocalDate = LocalDate.now(),
): Int? =
    runCatching {
        val bd = LocalDate.parse(isoDate)
        val next = nextBirthdayDate(isoDate, today) ?: return@runCatching null
        next.year - bd.year
    }.getOrNull()

/** Urgency buckets for the countdown pill. Mirrors iOS: today, within a week, later. */
internal enum class BirthdayUrgency { TODAY, SOON, LATER }

/** Days threshold below which a birthday is "soon" (amber-tinted pill). */
internal const val BIRTHDAY_SOON_DAYS = 7

/** Classifies a days-until-next-birthday count into an urgency bucket. */
internal fun birthdayUrgency(daysUntil: Int): BirthdayUrgency =
    when {
        daysUntil <= 0 -> BirthdayUrgency.TODAY
        daysUntil <= BIRTHDAY_SOON_DAYS -> BirthdayUrgency.SOON
        else -> BirthdayUrgency.LATER
    }

File: ios/FamilyApp/Features/Birthday/BirthdayDateUtils.swift
// Birthday date math.
import Foundation

/// Next occurrence of an ISO birthday on or after `today`; nil when malformed.
func nextBirthdayDate(_ isoDate: String, today: LocalDate = .today()) -> LocalDate? {
    guard let birthday = LocalDate(iso: isoDate) else { return nil }
    let thisYear = birthday.withYear(today.year)
    return thisYear < today ? birthday.withYear(today.year + 1) : thisYear
}

/// The age the person turns on their next birthday; nil when malformed.
func turnsAge(_ isoDate: String, today: LocalDate = .today()) -> Int? {
    guard let birthday = LocalDate(iso: isoDate),
          let next = nextBirthdayDate(isoDate, today: today) else { return nil }
    return next.year - birthday.year
}
